File: lib/services/flight_school_service.dart
import 'package:appwrite/appwrite.dart';
import 'package:flutter/cupertino.dart';
import 'package:season_planer/services/database_service.dart';
import 'package:season_planer/services/functions/flight_school_functions.dart';

import '../core/appwrite_config.dart';
import '../data/enums/event_role_enum.dart';
import '../data/enums/event_status_enum.dart';
import '../data/enums/event_user_status_enum.dart';
import '../data/models/admin_models/flight_school_model_flight_school_view.dart';
import '../data/models/admin_models/user_summary_flight_school_view.dart';
import '../data/models/event_model.dart';

class FlightSchoolService {
  final Client client = Client()
      .setEndpoint(AppwriteConfig().appwriteEnpoint)
      .setProject(AppwriteConfig().projectId)
      .setSelfSigned(status: true);

  late final flighSchoolFunctions = FlightSchoolFunctions(client);

  static final FlightSchoolService _instance = FlightSchoolService._internal();

  factory FlightSchoolService(){
    return _instance;
  }

  FlightSchoolService._internal();

  late final Databases _database;

  void init() {
    _database = Databases(client);
  }

  //TODO: Move remaining methods for FS

  //update Image
  //update Informations

  Future<bool> updateAdmins({required String flightSchoolId, required List<String> adminUserIds}) async {
    try {
      await _database.updateDocument(
        databaseId: AppwriteConfig().mainDatabaseId,
        collectionId: AppwriteConfig().flightSchoolsCollectionId,
        documentId: flightSchoolId,
        data: {
          'admin_users': adminUserIds,
        },
      );

      return true;
    } catch (e) {
      debugPrint("updateAdmins failed: $e");
      return false;
    }
  }


  Future<FlightSchoolModelFlightSchoolView?> getFlightSchool(String id) async {
    try {
      final fsDoc = await _fetchFlightSchoolDoc(id);
      final fs = fsDoc.data;

      final databaseId = (fs["database_id"] ?? "") as String;
      final teamAssignmentsId = (fs["team_assigments_events_id"] ?? "") as String;
      final eventsCollectionId = (fs["events_id"] ?? "") as String;

      final members = await _fetchMembers(id);
      final events = await _fetchEvents(
        flightSchoolId: id,
        databaseId: databaseId,
        eventsCollectionId: eventsCollectionId,
        teamAssignmentsId: teamAssignmentsId,
      );

      final settings =
          (fs["settings"] as Map<String, dynamic>?) ?? <String, dynamic>{};

      return FlightSchoolModelFlightSchoolView(
        id: fsDoc.$id,
        displayName: (fs["display_name"] ?? "") as String,
        displayShortName: (fs["display_short_name"]) as String,
        databaseId: databaseId,
        teamAssignmentsEventsCollectionId: teamAssignmentsId,
        eventsCollectionId: eventsCollectionId,
        auditLogsCollectionId: (fs["audit_logs_id"] ?? "") as String,
        logoLink: (fs["logo_link"] ?? "!") as String,
        logoId: (fs["logo_id"] ?? ""),
        adminUserIds: List<String>.from(fs["admin_users"] ?? const []),
        members: members,
        events: events,
        settings: settings,
      );
    } catch (e) {
      print("getFlightSchool (FS view) error: $e");
      return null;
    }
  }

  Future<dynamic> _fetchFlightSchoolDoc(String id) async {
    return _database.getDocument(
      databaseId: AppwriteConfig().mainDatabaseId,
      collectionId: AppwriteConfig().flightSchoolsCollectionId,
      documentId: id,
    );
  }

  Future<List<UserSummary>> _fetchMembers(String flightSchoolId) async {
    List<UserSummary> members = [];

    try {
      final fsFn = FlightSchoolFunctions(client);
      final result = await fsFn.getMembersWithAuth(
        flightSchoolId: flightSchoolId,
      );

      if (result is List) {
        members = result.map((m) {
          final mm = (m is Map) ? m : <String, dynamic>{};
          return UserSummary(
            id: (mm["userId"] ?? "").toString(),
            name: (mm["name"] ?? "").toString(),
            mail: (mm["email"] ?? "").toString(),
            phone: (mm["phone"] ?? "").toString(),
            membershipId: (mm["membership"]["id"]).toString(),
            roles: UserSummary.parseRoles(
              mm["membership"]?["roles"],
            ),
          );
        }).toList();
      } else {
        members = [];
      }
    } catch (e) {
      members = [];
      debugPrint("getMembersWithAuth failed: $e");
    }

    return members;
  }

  Future<List<Event>> _fetchEvents({
    required String flightSchoolId,
    required String databaseId,
    required String eventsCollectionId,
    required String teamAssignmentsId,
  }) async {
    List<Event> events = [];

    try {
      if (databaseId.isNotEmpty && eventsCollectionId.isNotEmpty) {
        final eventsResult = await _database.listDocuments(
          databaseId: databaseId,
          collectionId: eventsCollectionId,
          queries: [
            Query.orderDesc("start_time"),
            Query.limit(200),
          ],
        );

        final List<Event> eventList = [];

        for (final eventDoc in eventsResult.documents) {
          final eventData = eventDoc.data;

          final teamDocs = await _database.listDocuments(
            databaseId: databaseId,
            collectionId: teamAssignmentsId,
            queries: [
              Query.equal("events", eventDoc.$id),
            ],
          );

          final teamMembers = teamDocs.documents
              .map((d) => TeamMember.fromMap(d.data))
              .toList();

          final firstAssignment =
          teamDocs.documents.isNotEmpty ? teamDocs.documents.first.data : null;

          eventList.add(
            Event(
              id: eventDoc.$id,
              flightSchoolId: flightSchoolId,
              identifier: (eventData["identifier"] ?? "test").toString(),
              status: EventStatusEnum.values.byName(eventData["status"]),
              startTime: DateTime.parse(eventData["start_time"]),
              endTime: DateTime.parse(eventData["end_time"]),
              displayName: (eventData["display_name"] ?? "test").toString(),
              team: teamMembers,
              notes: eventData["notes"],
              location: eventData["location"],
              role: firstAssignment != null
                  ? EventRoleEnum.values.byName(firstAssignment["role"])
                  : EventRoleEnum.values.first,
              assignmentStatus: firstAssignment != null
                  ? EventUserStatusEnum.values.byName(firstAssignment["status"])
                  : EventUserStatusEnum.values.first,
            ),
          );
        }

        events = eventList;
      }
    } catch (e) {
      print("events loading error: $e");
      events = [];
    }

    return events;
  }



  ///Event Handling
  ///

  /// Member Handling
Future<bool> removeMemberOfFlightSchool(String membershipId) async {
    try{
      _database.deleteDocument(
          databaseId: AppwriteConfig().mainDatabaseId,
          collectionId: AppwriteConfig().membershipsId,
          documentId: membershipId,
      );
      return false;
    }catch(e){
      print(e);
      return false;
    }
}

Future<bool> updateRolesInMemberOfFlightSchool(String membershipId,List roles) async {
    try{
      _database.updateDocument(
          databaseId: AppwriteConfig().mainDatabaseId,
          collectionId: AppwriteConfig().membershipsId,
          documentId: membershipId,
          data: {
               "roles": roles
          }
      );
      return true;
    }catch(e){
      print(e);
      return false;
    }
  }

  Future<void> inviteMember({
    required String flightSchoolId,
    required String email,
  }) async {
    final roleNames = [""];

    final res = await flighSchoolFunctions.inviteUserToFlightSchool(
      flightSchoolId: flightSchoolId,
      userMail: email,
      roles: roleNames,
    );

    if (res["ok"] != true) {
      throw Exception("Invite failed: ${res.toString()}");
    }
  }



}

File: appwrite_functions/flightSchoolServices/src/services/flightSchool.service.js
import { ID, Query } from "node-appwrite";
import { createAppwrite } from "../../src/appwrite/client.js";

const MAIN_DATABASE_ID = process.env.MAIN_DATABASE_ID;
const FLIGHT_SCHOOLS_COLLECTION_ID = process.env.FLIGHT_SCHOOLS_COLLECTION_ID;
const MEMBERSHIPS_COLLECTION_ID = process.env.MEMBERSHIPS_COLLECTION_ID;
const EVENTS_COLLECTION_ID = process.env.EVENTS_COLLECTION_ID;
const TEAM_ASSIGNMENTS_COLLECTION_ID = process.env.TEAM_ASSIGNMENTS_COLLECTION_ID;

const { users, databases } = createAppwrite();

export async function getFlightSchoolAdminView({ flightSchoolId, context } = {}) {
  logInfo(context, "getFlightSchoolAdminView:start");

  logJson(context, "env", {
    MAIN_DATABASE_ID,
    FLIGHT_SCHOOLS_COLLECTION_ID,
    MEMBERSHIPS_COLLECTION_ID,
    EVENTS_COLLECTION_ID,
    TEAM_ASSIGNMENTS_COLLECTION_ID,
    flightSchoolId,
  });

  if (!flightSchoolId) {
    logErrorMessage(context, "FLIGHT_SCHOOL_ID_REQUIRED");
    return { ok: false, error: "FLIGHT_SCHOOL_ID_REQUIRED" };
  }

  try {
    requireEnv(MAIN_DATABASE_ID, "MAIN_DATABASE_ID");
    requireEnv(FLIGHT_SCHOOLS_COLLECTION_ID, "FLIGHT_SCHOOLS_COLLECTION_ID");

    const fsDoc = await databases.getDocument(
      MAIN_DATABASE_ID,
      FLIGHT_SCHOOLS_COLLECTION_ID,
      flightSchoolId
    );

    logJson(context, "fsDoc_meta", {
      id: fsDoc?.$id ?? null,
      keys: Object.keys(fsDoc ?? {}),
    });

    const fs = getDocumentData(fsDoc);

    const members = await getMembersWithAuth({ flightSchoolId, context });
    const events = await getFlightSchoolEvents({ flightSchoolId, context });

    const result = {
      ok: true,
      flightSchool: {
        id: fsDoc.$id,
        displayName: fs.display_name ?? "",
        displayShortName: fs.display_short_name ?? "",
        databaseId: fs.database_id ?? "",
        teamAssignmentsEventsCollectionId: fs.team_assigments_events_id ?? "",
        eventsCollectionId: fs.events_id ?? "",
        auditLogsCollectionId: fs.audit_logs_id ?? "",
        logoLink: fs.logo_link ?? "",
        logoId: fs.logo_id ?? "",
        adminUserIds: Array.isArray(fs.admin_users) ? fs.admin_users : [],
        settings: isPlainObject(fs.settings) ? fs.settings : {},
        members,
        events,
      },
    };

    logJson(context, "getFlightSchoolAdminView:success", {
      flightSchoolId: result.flightSchool.id,
      membersCount: result.flightSchool.members.length,
      eventsCount: result.flightSchool.events.length,
    });

    return result;
  } catch (error) {
    logError(context, "getFlightSchoolAdminView:error", error);
    return { ok: false, error: "FLIGHT_SCHOOL_LOAD_FAILED" };
  }
}

export async function getMembersWithAuth({ flightSchoolId, context } = {}) {
  logJson(context, "getMembersWithAuth:start", {
    flightSchoolId,
    MAIN_DATABASE_ID,
    MEMBERSHIPS_COLLECTION_ID,
  });

  try {
    requireEnv(MAIN_DATABASE_ID, "MAIN_DATABASE_ID");
    requireEnv(MEMBERSHIPS_COLLECTION_ID, "MEMBERSHIPS_COLLECTION_ID");

    const membershipDocs = await databases.listDocuments(
      MAIN_DATABASE_ID,
      MEMBERSHIPS_COLLECTION_ID,
      [
        Query.equal("flightSchools", [flightSchoolId]),
        Query.limit(500),
      ]
    );

    const memberships = membershipDocs?.documents ?? [];

    logJson(context, "getMembersWithAuth:listDocuments:result", {
      total: memberships.length,
    });

    const result = await Promise.all(
      memberships.map(async (membershipDoc, index) => {
        const membership = getDocumentData(membershipDoc);
        const userId = extractUserIdFromMembership(membership);

        logJson(context, "getMembersWithAuth:membership_item", {
          index,
          membershipId: membershipDoc?.$id ?? null,
          userId,
          roles: Array.isArray(membership.roles) ? membership.roles : [],
          status: membership.status ?? "",
        });

        let authUser = null;

        if (userId) {
          try {
            authUser = await users.get(userId);
          } catch (error) {
            logError(context, "getMembersWithAuth:users.get:error", error);
            authUser = null;
          }
        }

        return {
          userId: userId ?? "",
          name: authUser?.name ?? "",
          email: authUser?.email ?? "",
          phone: authUser?.phone ?? "",
          membership: {
            id: membershipDoc.$id,
            roles: Array.isArray(membership.roles) ? membership.roles : [],
            status: membership.status ?? "",
          },
        };
      })
    );

    logJson(context, "getMembersWithAuth:success", {
      count: result.length,
    });

    return result;
  } catch (error) {
    logError(context, "getMembersWithAuth:error", error);
    return [];
  }
}

export async function inviteUserToFlightSchool({
  flightSchoolId,
  userMail,
  roles,
  context,
} = {}) {
  logJson(context, "inviteUserToFlightSchool:start", {
    flightSchoolId,
    userMail,
    roles,
  });

  if (!flightSchoolId) {
    return { ok: false, error: "FLIGHT_SCHOOL_ID_REQUIRED" };
  }

  if (!userMail) {
    return { ok: false, error: "USER_MAIL_REQUIRED" };
  }

  try {
    const userList = await users.list([
      Query.equal("email", [userMail]),
      Query.limit(1),
    ]);

    const authUser = userList?.users?.[0];

    if (!authUser) {
      return { ok: false, error: "USER_NOT_FOUND" };
    }

    const existingMemberships = await databases.listDocuments(
      MAIN_DATABASE_ID,
      MEMBERSHIPS_COLLECTION_ID,
      [
        Query.equal("users", [authUser.$id]),
        Query.equal("flightSchools", [flightSchoolId]),
        Query.limit(1),
      ]
    );

    if ((existingMemberships?.documents ?? []).length > 0) {
      return { ok: false, error: "MEMBERSHIP_ALREADY_EXISTS" };
    }

    const created = await databases.createDocument(
      MAIN_DATABASE_ID,
      MEMBERSHIPS_COLLECTION_ID,
      ID.unique(),
      {
        users: authUser.$id,
        flightSchools: flightSchoolId,
        roles: Array.isArray(roles) ? roles : [],
        status: "pending",
      }
    );

    return {
      ok: true,
      membershipId: created.$id,
    };
  } catch (error) {
    logError(context, "inviteUserToFlightSchool:error", error);
    return { ok: false, error: "INVITE_FAILED" };
  }
}

export async function createNewEvent({context,
  event
} = {}){
  try{
  const created = await databases.createDocument(
    MAIN_DATABASE_ID,
    EVENTS_COLLECTION_ID,
    ID.unique(),
    {
      flight_school_id: event.flightSchoolId,
      identifier: event.identifier ,
      status: event.status,
      display_name: event.displayName,
      start_time: event.startTime,
      end_time: event.endTime,
      notes: event.notes,
      location: event.location
    }
  );

  for (const tm of event.team ?? []) {
      context?.log?.(`team item: ${JSON.stringify(tm)}`);

      await databases.createDocument(
        MAIN_DATABASE_ID,
        TEAM_ASSIGNMENTS_COLLECTION_ID,
        ID.unique(),
        {
          user_id: tm.user_id ?? "",
          role: tm.role ?? "",
          status: tm.status ?? "",
          events: created.$id,
        }
      );
    }

  return{
    ok: true,
    eventID: created.$id
  }

  }catch(error){
    logError(context, "createNewEvent:error", error);
    return { ok: false, error: "CREATE_EVENT_FAILED" };
  }
 
}

export async function updateFlightSchoolAdmins({
  flightSchoolId,
  adminUserIds,
  context,
} = {}) {
  logJson(context, "updateFlightSchoolAdmins:start", {
    flightSchoolId,
    adminUserIds,
  });

  if (!flightSchoolId) {
    return { ok: false, error: "FLIGHT_SCHOOL_ID_REQUIRED" };
  }

  try {
    const updated = await databases.updateDocument(
      MAIN_DATABASE_ID,
      FLIGHT_SCHOOLS_COLLECTION_ID,
      flightSchoolId,
      {
        admin_users: Array.isArray(adminUserIds) ? adminUserIds : [],
      }
    );

    return {
      ok: true,
      flightSchoolId: updated.$id,
    };
  } catch (error) {
    logError(context, "updateFlightSchoolAdmins:error", error);
    return { ok: false, error: "UPDATE_ADMINS_FAILED" };
  }
}

export async function removeMemberFromFlightSchool({ membershipId, context } = {}) {
  logJson(context, "removeMemberFromFlightSchool:start", { membershipId });

  if (!membershipId) {
    return { ok: false, error: "MEMBERSHIP_ID_REQUIRED" };
  }

  try {
    await databases.deleteDocument(
      MAIN_DATABASE_ID,
      MEMBERSHIPS_COLLECTION_ID,
      membershipId
    );

    return { ok: true };
  } catch (error) {
    logError(context, "removeMemberFromFlightSchool:error", error);
    return { ok: false, error: "REMOVE_MEMBER_FAILED" };
  }
}

export async function updateMemberRoles({ membershipId, roles, context } = {}) {
  logJson(context, "updateMemberRoles:start", {
    membershipId,
    roles,
  });

  if (!membershipId) {
    return { ok: false, error: "MEMBERSHIP_ID_REQUIRED" };
  }

  try {
    const updated = await databases.updateDocument(
      MAIN_DATABASE_ID,
      MEMBERSHIPS_COLLECTION_ID,
      membershipId,
      {
        roles: Array.isArray(roles) ? roles : [],
      }
    );

    return {
      ok: true,
      membershipId: updated.$id,
    };
  } catch (error) {
    logError(context, "updateMemberRoles:error", error);
    return { ok: false, error: "UPDATE_MEMBER_ROLES_FAILED" };
  }
}

export async function updateEvent({ event, context } = {}) {
  try {
    if (!event) throw new Error("EVENT_REQUIRED");

    const eventId = event.id ?? event.$id;
    if (!eventId) throw new Error("EVENT_ID_REQUIRED");

    requireEnv(MAIN_DATABASE_ID, "MAIN_DATABASE_ID");
    requireEnv(EVENTS_COLLECTION_ID, "EVENTS_COLLECTION_ID");
    requireEnv(TEAM_ASSIGNMENTS_COLLECTION_ID, "TEAM_ASSIGNMENTS_COLLECTION_ID");

    logJson(context, "updateEvent:start", {
      eventId,
      flightSchoolId: event.flightSchoolId,
      teamCount: Array.isArray(event.team) ? event.team.length : 0,
    });

    const updated = await databases.updateDocument(
      MAIN_DATABASE_ID,
      EVENTS_COLLECTION_ID,
      eventId,
      {
        identifier: event.identifier ?? "",
        status: event.status ?? "",
        start_time: event.startTime ?? null,
        end_time: event.endTime ?? null,
        notes: event.notes ?? "",
        display_name: event.displayName ?? "",
        location: event.location ?? "",
        flight_school_id: event.flightSchoolId ?? "",
      }
    );

    const existingTeamResult = await databases.listDocuments(
      MAIN_DATABASE_ID,
      TEAM_ASSIGNMENTS_COLLECTION_ID,
      [
        Query.equal("events", [eventId]),
        Query.limit(500),
      ]
    );

    const existingTeamDocs = existingTeamResult?.documents ?? [];
    const existingById = new Map(existingTeamDocs.map((doc) => [doc.$id, doc]));

    const incomingTeam = Array.isArray(event.team) ? event.team : [];
    const touchedIds = new Set();

    for (const tm of incomingTeam) {
      const assignmentId = tm.id ?? tm.$id ?? "";

      const data = {
        events: eventId,
        user_id: tm.userId ?? tm.user_id ?? "",
        role: tm.role ?? "",
        status: tm.status ?? "",
      };

      if (assignmentId && existingById.has(assignmentId)) {
        await databases.updateDocument(
          MAIN_DATABASE_ID,
          TEAM_ASSIGNMENTS_COLLECTION_ID,
          assignmentId,
          data
        );

        touchedIds.add(assignmentId);

        logJson(context, "updateEvent:team:update", {
          assignmentId,
          role: data.role,
          status: data.status,
          userId: data.user_id,
        });
      } else {
        const createdTeam = await databases.createDocument(
          MAIN_DATABASE_ID,
          TEAM_ASSIGNMENTS_COLLECTION_ID,
          ID.unique(),
          data
        );

        touchedIds.add(createdTeam.$id);

        logJson(context, "updateEvent:team:create", {
          assignmentId: createdTeam.$id,
          role: data.role,
          status: data.status,
          userId: data.user_id,
        });
      }
    }

    for (const existingDoc of existingTeamDocs) {
      if (!touchedIds.has(existingDoc.$id)) {
        await databases.updateDocument(
          MAIN_DATABASE_ID,
          TEAM_ASSIGNMENTS_COLLECTION_ID,
          existingDoc.$id,
          {
            status: "removed",
          }
        );

        logJson(context, "updateEvent:team:mark_removed", {
          assignmentId: existingDoc.$id,
        });
      }
    }

    logJson(context, "updateEvent:success", {
      eventId: updated.$id,
      incomingTeamCount: incomingTeam.length,
      existingTeamCount: existingTeamDocs.length,
    });

    return {
      ok: true,
      eventId: updated.$id,
    };
  } catch (error) {
    logError(context, "updateEvent:error", error);
    return { ok: false, error: "UPDATE_EVENT_FAILED" };
  }
}

export async function getFlightSchoolEvents({ flightSchoolId, context } = {}) {
  logJson(context, "getFlightSchoolEvents:start", {
    flightSchoolId,
    MAIN_DATABASE_ID,
    EVENTS_COLLECTION_ID,
    TEAM_ASSIGNMENTS_COLLECTION_ID,
  });

  if (!flightSchoolId) {
    return [];
  }

  try {
    requireEnv(MAIN_DATABASE_ID, "MAIN_DATABASE_ID");
    requireEnv(EVENTS_COLLECTION_ID, "EVENTS_COLLECTION_ID");
    requireEnv(TEAM_ASSIGNMENTS_COLLECTION_ID, "TEAM_ASSIGNMENTS_COLLECTION_ID");

    const eventsResult = await databases.listDocuments(
      MAIN_DATABASE_ID,
      EVENTS_COLLECTION_ID,
      [
        Query.equal("flight_school_id", [flightSchoolId]),
        Query.orderDesc("start_time"),
        Query.limit(200),
      ]
    );

    const eventDocs = eventsResult?.documents ?? [];

    logJson(context, "getFlightSchoolEvents:eventDocs", {
      count: eventDocs.length,
    });

    const events = await Promise.all(
      eventDocs.map(async (eventDoc) => {
        const eventData = getDocumentData(eventDoc);

        const teamDocs = await databases.listDocuments(
          MAIN_DATABASE_ID,
          TEAM_ASSIGNMENTS_COLLECTION_ID,
          [
            Query.equal("events", [eventDoc.$id]),
            Query.notEqual("status", "removed"),
            Query.limit(200),
          ]
        );

        const team = (teamDocs?.documents ?? []).map((doc) => {
          const d = getDocumentData(doc);
          return {
            id: doc.$id,
            userId: d.user_id ?? "",
            role: d.role ?? "",
            status: d.status ?? "",
            name: d.name ?? "",
          };
        });

        return {
          id: eventDoc.$id,
          flightSchoolId,
          identifier: eventData.identifier ?? "",
          status: eventData.status ?? "",
          startTime: eventData.start_time ?? null,
          endTime: eventData.end_time ?? null,
          displayName: eventData.display_name ?? "",
          location: eventData.location ?? "",
          notes: eventData.notes ?? "",
          team,
        };
      })
    );

    logJson(context, "getFlightSchoolEvents:success", {
      count: events.length,
    });

    return events;
  } catch (error) {
    logError(context, "getFlightSchoolEvents:error", error);
    return [];
  }
}

function getDocumentData(doc) {
  return doc?.data ?? doc ?? {};
}

function extractUserIdFromMembership(membership) {
  const usersField = membership?.users;

  if (!usersField) return null;
  if (typeof usersField === "string") return usersField;
  if (typeof usersField === "object" && usersField.$id) return usersField.$id;

  return null;
}

function isPlainObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function requireEnv(value, name) {
  if (!value || String(value).trim() === "") {
    throw new Error(`${name}_MISSING`);
  }
}

function logInfo(context, message) {
  if (context?.log && typeof context.log === "function") {
    context.log(message);
    return;
  }
  console.log(message);
}

function logErrorMessage(context, message) {
  if (context?.error && typeof context.error === "function") {
    context.error(message);
    return;
  }
  console.error(message);
}

function logError(context, message, error) {
  const payload = {
    message,
    errorMessage: error?.message ?? "UNKNOWN_ERROR",
    errorName: error?.name ?? null,
    errorCode: error?.code ?? null,
    errorType: error?.type ?? null,
    errorResponse: error?.response ?? null,
  };

  if (context?.error && typeof context.error === "function") {
    context.error(JSON.stringify(payload));
    return;
  }

  console.error(payload);
}

function logJson(context, label, value) {
  const payload = `${label}: ${JSON.stringify(value)}`;

  if (context?.log && typeof context.log === "function") {
    context.log(payload);
    return;
  }

  console.log(payload);
}